core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -0,0 +1,264 @@
+
+    val updatedAcls: Set[Acl] = getAcls(resource) ++ acls
+    val path: String = toResourcePath(resource)
+
+    if (ZkUtils.pathExists(zkClient, path))
+      ZkUtils.updatePersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(updatedAcls)))
+    else
+      ZkUtils.createPersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(updatedAcls)))
+
+    inWriteLock(lock) {
+      aclCache.put(resource, updatedAcls)
+    }
+
+    updateAclChangedFlag(resource)
+  }
+
+  override def removeAcls(acls: Set[Acl], resource: Resource): Boolean = {
+    if (!ZkUtils.pathExists(zkClient, toResourcePath(resource)))
+      return false
+
+    val existingAcls: Set[Acl] = getAcls(resource)
+    val filteredAcls: Set[Acl] = existingAcls.filter((acl: Acl) => !acls.contains(acl))
+    if (existingAcls.equals(filteredAcls))
+      return false
+
+    val path: String = toResourcePath(resource)
+    ZkUtils.updatePersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(filteredAcls)))
+
+    inWriteLock(lock) {
+      aclCache.put(resource, filteredAcls)
+    }
+
+    updateAclChangedFlag(resource)
+
+    true
+  }
+
+  override def removeAcls(resource: Resource): Boolean = {
+    if (ZkUtils.pathExists(zkClient, toResourcePath(resource))) {
+      ZkUtils.deletePath(zkClient, toResourcePath(resource))
+      inWriteLock(lock) {
+        aclCache.remove(resource)
+      }
+      updateAclChangedFlag(resource)
+      true
+    } else false
+  }
+
+  override def getAcls(resource: Resource): Set[Acl] = {
+    inReadLock(lock) {
+      if (aclCache.contains(resource))
+        return aclCache.get(resource).get
+    }
+
+    val aclJson: Option[String] = ZkUtils.readDataMaybeNull(zkClient, toResourcePath(resource))._1
+    val acls: Set[Acl] = aclJson map ((x: String) => Acl.fromJson(x)) getOrElse Set.empty
+
+    inWriteLock(lock) {
+      aclCache.put(resource, acls)
+    }
+    acls
+  }
+
+  override def getAcls(principal: KafkaPrincipal): Set[Acl] = {
+    val aclSets = aclCache.values
+    var acls = Set.empty[Acl]
+    for(aclSet: Set[Acl]  <- aclSets) {
+      acls ++= aclSet.filter(acl => acl.principal.equals(principal))
+    }
+    acls.toSet[Acl]
+  }
+
+  def loadCache: Set[Acl] = {
+    var acls = Set.empty[Acl]
+    val resourceTypes = ZkUtils.getChildren(zkClient, SimpleAclAuthorizer.AclZkPath)
+    for (rType <- resourceTypes) {
+      val resourceType = ResourceType.fromString(rType)
+      for (resources <- ZkUtils.getChildren(zkClient, resourceType.name)) {
+        for (resourceName <- resources)
+          acls ++= getAcls(Resource(resourceType, resourceName.toString))
+      }
+    }
+    acls
+  }
+
+
+  def toResourcePath(resource: Resource): String = {
+    SimpleAclAuthorizer.AclZkPath + "/" + resource.resourceType + "/" + resource.name
+  }
+
+  def updateAclChangedFlag(resource: Resource): Unit = {
+      ZkUtils.updatePersistentPath(zkClient, SimpleAclAuthorizer.AclChangedZkPath, resource.toString)
```

Comment:
Parth,

ZK watchers are actually one-timer watchers (http://zookeeper.apache.org/doc/r3.4.1/zookeeperProgrammers.html#sc_WatchRememberThese). When a watcher fires, the client has to register the watcher again (typically through a read) to get notification for future changes. Between the time that a watcher fires and the client does a read, multiple changes could have happened. The read will return the latest value and leave a watcher there. So, in general, there is no guarantee that every change triggers the firing of a watcher. In particular, if you update the value of a ZK path multiple times in between, only the latest value will be seen by the reader.

One way to capture all changes is to follow how we propagate the config changes (ConfigChangeListener). Whenever we change an ACL, we write the ACL under /kafka-acl and also write a sequential ZK node under /kafka-acl-changed. The value of the sequential ZK node indicates the resource for which the ACL has changed. The sequential ZK nodes are guaranteed to be unique and have a number suffix that's ever growing. The ACL manager registers a child change listener on /kafka-acl-changed and remembers the last number suffix that it has processed. When the watcher fires, the ACL manager reads all child nodes under /kafka-acl-changed (the read may pick up multiple ACL changes) and process new nodes after the last remembered number suffix. The processing will involve reading the latest ACL associated with the corresponding resource. Since every sequential node is unique, the ACL manager won't miss any ACL change. 

On initialization, the ACL manager will first register the watcher and read all existing ACLs.

Finally, we need a way to garbage collect old sequential nodes. This can be done by just removing sequential nodes that are say, more than 15 mins old (assuming every broker would have picked up the ACL changes by then).